JobScheduler.js

```diff
@@ -1,84 +1,215 @@
+      return maintenanceManager
+        .getMaintenaceInfo()
+        .then(maintenanceInfo => {
+          if (maintenanceInfo === null) {
+            logger.info('+-> System is not in maintenance');
+            if (this.intervalTimer) {
+              clearInterval(this.intervalTimer);
+            }
+            return resolve();
+          } else {
+            logger.info('+-> System is in maintenance:', maintenanceInfo);
+            const currTime = moment();
+            const maintenanceStartTime = _.get(maintenanceInfo, 'createdAt');
+            if (maintenanceStartTime && currTime.diff(maintenanceStartTime) > config.scheduler.maintenance_mode_time_out) {
+              logger.warn(`System in maintenance since ${maintenanceInfo.createdAt}. Exceeds configured maintenance timeout. Flagging the current maintenance window as aborted.`);
+              const success = resolve;
+              return maintenanceManager
+                .updateMaintenace(`System in maintenance beyond configured timeout time ${config.scheduler.maintenance_mode_time_out/1000/60} (mins)`,
+                  CONST.OPERATION.ABORTED,
+                  CONST.SYSTEM_USER)
+                .then(success)
+                .catch((err) => {
+                  logger.error('error occurred while updating maintenance info', err);
+                  reject(err);
+                })
+                .finally(() => {
+                  if (this.intervalTimer) {
+                    clearInterval(this.intervalTimer);
+                  }
+                });
+            }
+            if (this.intervalTimer === undefined) {
+              logger.info(`Poll maintenance status once every - ${config.scheduler.maintenance_check_interval} (ms)`);
+              this.intervalTimer = setInterval(() => checkMaintenanceStatus.call(this, resolve, reject),
+                config.scheduler.maintenance_check_interval);
+              return;
+            }
+          }
+        })
+        .catch(err => reject(err));
+    };
+    return new Promise((resolve, reject) => {
+      checkMaintenanceStatus.call(this, resolve, reject);
+    });
   }
-});
\ No newline at end of file
+
+  ensureSystemNotInMainenanceThenInitMaster() {
+    logger.info('checking if system in maintenance...');
+    return this
+      .pollMaintenanceStatus()
+      .then(() => this.initMaster())
+      .catch(err => {
+        logger.error('Error occurred while checking maintenance / init. Scheduler will exit shortly..', err);
+        setTimeout(() => process.exit(CONST.ERR_CODES.INTERNAL_ERROR), CONST.JOB_SCHEDULER.SHUTDOWN_WAIT_TIME);
+      });
+  }
+
+  processUnhandledRejection(reason, p) {
+    logger.error(`Unhandled Rejection in - ${this.workerType} - at:`, p, 'reason:', reason);
+    if (reason && reason instanceof errors.DBUnavailable) {
+      logger.error('DB unavailable. shutting down app');
+      this.shutDown();
+    } else {
+      logger.error('Unhandled Rejection at:', p, 'reason:', reason);
```

Comment:
No. Ideally all code must handle the rejections / errors as part of their code only. And we also have a global error handler in middleware.js which takes care of any unhandled error globally and responds back. 
This is put in place for two things:
1. If there are some places (which run in background) in code which are not handling errors (which necessarily need not be fatal) are logged, so that they can be fixed at source and rejections are handled.
2. From a background process like (DBConnectionManager) when we want to throw an exception (very deterministic error) which should result in shutdown, then a shutdown is to be initiated. In future if there is another deterministic error which must result in shutdown then the same error type must be added in the above. 

Shutting down on any unhandled rejection is not required.

Reply:
Ok fine.